src/badguy/ghoul.cpp

```diff
@@ -15,202 +15,339 @@
+      set_action(new_dir == Direction::LEFT ? "left-up" : "right-up");
+      m_dir = new_dir;
+    }
+    if (chase) {
+      set_state(CHASING_UP);
+    } else if (m_physic.get_velocity_y() > DOWN_VELOCITY) {
+      if (get_pos().y > m_home_pos.y + VERT_OFFSET) {
+        set_state(ROAMING_ACCEL1);
+      } else {
+        set_state(ROAMING_DOWN);
+      }
+    }
+    break;
+  case CHASING_UP:
+    update_speed(dist);
+    if (dir_changed) {
+      set_action(new_dir == Direction::LEFT ? "left-up" : "right-up");
+      m_dir = new_dir;
+    }
+    if (!chase) {
+      m_home_pos = get_pos();
+      set_state(ROAMING_UP);
+    } else if (m_physic.get_velocity_y() > DOWN_VELOCITY) {
+      if (dist.y < -VERT_OFFSET) {
+        set_state(CHASING_ACCEL1);
+      } else {
+        set_state(CHASING_DOWN);
+      }
+    }
+    break;
+  case STUNNED:
+    if (m_sprite->animation_done()) {
+      set_state(INVISIBLE);
+    }
+    break;
+  case INVISIBLE:
+    if (m_respawn_timer.check()) {
+      set_state(RECOVERING);    
+    }
+    break;
+  case RECOVERING:
+    if (m_sprite->animation_done()) {
+      set_state(ROAMING_DOWN);
+    }
+    break;
+  }
+
+  m_dir = new_dir;
 }
 
 void
-Ghoul::finish_construction()
+Ghoul::update_speed(const Vector& dist)
 {
-  if (get_walker() && get_walker()->is_running()) {
-    m_mystate = STATE_PATHMOVING_TRACK;
+  const float vx = m_physic.get_velocity_x();
+  if (vx >= -HORZ_SPEED && vx <= HORZ_SPEED) {
+  	m_physic.set_acceleration_x(0.0f);
+  	const float vy = dist.y < 0.0f ? (UP_VELOCITY + DOWN_VELOCITY) / 2.0f : DOWN_VELOCITY;
+    const float t = dist.y / vy;
+    if (t * HORZ_SPEED > std::abs(dist.x)) {
+      m_physic.set_velocity_x(dist.x / t);
+    } else {
+      m_physic.set_velocity_x(dist.x < 0.0f ? -HORZ_SPEED : HORZ_SPEED);
+    }
   }
 }
 
 void
-Ghoul::activate()
+Ghoul::draw(DrawingContext& context)
 {
-  if (Editor::is_active())
-    return;
+  if (m_state != INVISIBLE) {
+    BadGuy::draw(context);
+  }
 }
 
 void
-Ghoul::deactivate()
+Ghoul::horizontal_thrust()
 {
-  switch (m_mystate) {
-    case STATE_TRACKING:
-      m_mystate = STATE_IDLE;
-      break;
-    default:
-      break;
+  const float vy = (UP_VELOCITY + DOWN_VELOCITY) / 2.0f;
+  const Vector dist = to_target();
+  const float t = dist.y / vy;
+  if (t * (std::abs(m_physic.get_velocity_x()) + HORZ_SPEED) / 2.0f > std::abs(dist.x) || dist.y > 0)
+    return; //no need for acceleration
+
+  const float a = dist.x > 0.0f ? -HORZ_ACCELERATION : HORZ_ACCELERATION;
+  m_physic.set_acceleration_x(a);
+  const float vx = m_physic.get_velocity_x();
+  const float vx_diff = HORZ_SPEED * 9.0f;
+  if (t == 0.0f) {
+    m_physic.set_velocity_x(dist.x > 0.0f ? vx - vx_diff : vx + vx_diff);
+  } else {
+    const float vx_needed = dist.x / t - a * t / 2.0f;
+    m_physic.set_velocity_x(std::clamp(vx_needed, vx - vx_diff, vx + vx_diff));
   }
 }
 
 void
-Ghoul::active_update(float dt_sec)
+Ghoul::start_roaming_deccel()

```

Comment:
Shouldn't this be `decel` (typo)?

Reply:
Screw your arbitrary gramar rules, I'm an engineer and not a gramar expert :P